apps/anomalist/detectors.py

```diff
@@ -0,0 +1,193 @@
+from typing import Dict, List, Literal, Optional
+
+import numpy as np
+import pandas as pd
+import plotly.express as px
+import structlog
+from owid.datautils.dataframes import map_series
+
+from etl import grapher_model as gm
+from etl.data_helpers.misc import bard
+
+log = structlog.get_logger()
+
+# Name of index columns for dataframe.
+INDEX_COLUMNS = ["entity_id", "year"]
+
+# Define anomaly types.
+ANOMALY_TYPE = Literal["upgrade_change", "time_change", "upgrade_missing"]
+
+
+def estimate_bard_epsilon(series: pd.Series) -> float:
+    # Make all values positive, and ignore zeros.
+    positive_values = abs(series.dropna())
+    # Ignore zeros, since they can lead to epsilon being zero, hence allowing division by zero in BARD.
+    positive_values = positive_values.loc[positive_values > 0]
+    # Estimate epsilon as the absolute range of values divided by 10.
+    # eps = (positive_values.max() - positive_values.min()) / 10
+    # Instead of just taking maximum and minimum, take 95th percentile and 5th percentile.
+    eps = (positive_values.quantile(0.95) - positive_values.quantile(0.05)) / 10
+
+    return eps
+
+
+def get_long_format_score_df(df_score: pd.DataFrame) -> pd.DataFrame:
+    # Create a reduced score dataframe.
+    df_score_long = df_score.melt(
+        id_vars=["entity_id", "year"], var_name="variable_id", value_name="anomaly_score"
+    ).fillna(0)
+    # For now, keep only the latest year affected for each country-indicator.
+    df_score_long = (
+        df_score_long.sort_values("anomaly_score", ascending=False)
+        .drop_duplicates(subset=["variable_id", "entity_id"], keep="first")
+        .reset_index(drop=True)
+    )
+
+    return df_score_long
+
+
+class AnomalyDetector:
+    anomaly_type: str
+
+    def __init__(
+        self,
+        variable_ids: List[int],
```

Comment:
Passing data to `__init__` and accessing them via `self.` can get tricky. An alternative is to keep it stateless and pass only hyperparameters and config to `__init__` and actual data to `get_score_df` method. This would make it look like this
```
def get_score_df(
        self, df: pd.DataFrame, variables: list[gm.Variable], variable_mapping: dict[int, int]
    ) -> pd.DataFrame:
```

It makes it clear what data is it using and prevents unexpected memory leaks (when you don't delete the object). Scikit-learn uses that pattern too in `estimator.fit(X, y)` and `estimator.predict(X)`.